### PR TITLE
Add `group_mc_cv()`

### PR DESCRIPTION
**Pull request by mikemahoney218**
This PR addresses #207 by adding a new function, `group_mc_cv()`, which can be used to create grouped Monte Carlo cross-validation resamples. Note that the user-facing function doesn't have a `balance` argument, because it doesn't make sense to balance anything other than the proportion of data assigned to the assessment fold.

This PR _also_ re-organizes code so that grouped functions now live in the same file as their ungrouped variants. I _believe_ we were agreed that would make the most sense going forward, but happy to undo if not.

``` r
library(rsample)
data(ames, package = "modeldata")
set.seed(123)
group_mc_cv(ames, group = Neighborhood, times = 5)
#> # Grouped Monte Carlo cross-validation (0.75/0.25) with 5 resamples  
#> # A tibble: 5 × 2
#>   splits             id       
#>   <list>             <chr>    
#> 1 <split [2182/748]> Resample1
#> 2 <split [2196/734]> Resample2
#> 3 <split [1980/950]> Resample3
#> 4 <split [2056/874]> Resample4
#> 5 <split [2178/752]> Resample5
```

<sup>Created on 2022-06-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### CODE REVIEW

**Comment by mikemahoney218** [comment]
I _think_ that if we like how this is implemented, then we can also quickly add `group_validation_split()`, `group_initial_split()`, and `group_bootstraps()` (slightly harder, but only slightly). That's everything mentioned in #207, although I think we also want to implement stratification before considering that "done".

File: R/make_groups.R [hunk 1]
```diff
@@ -10,24 +10,26 @@
...
 #' @param ... Arguments passed to balance functions.
 #'
 #' @keywords internal
 make_groups <- function(data,
                         group,
                         v,
-                        balance = c("groups", "observations"),
+                        balance = c("groups", "observations", "prop"),
                         ...) {
+  rlang::check_dots_used(call = rlang::caller_env())
```
Comment (hfrick) [inline R/make_groups.R:22]: nice!

**Review by hfrick — APPROVED** [review]

**Review by juliasilge — COMMENTED** [review]
This is looking great! 🙌  I have just a couple of comments, one of which is about an existing example.

File: R/vfold.R [hunk 2]
```diff
@@ -140,6 +140,96 @@ vfold_splits <- function(data, v = 10, strata = NULL, breaks = 4, pool = 0.1) {
...
+#' data(Sacramento, package = "modeldata")
+#'
+#' set.seed(123)
+#' group_vfold_cv(Sacramento, group = city, v = 5)
+#' group_vfold_cv(
+#'   Sacramento,
+#'   group = city,
+#'   v = 5,
+#'   balance = "observations"
+#' )
```
[the remark this answers is not in the document]
Comment (mikemahoney218) [reply R/vfold.R:176]: Whoops! Swapped to `ames`:

``` r
library(rsample)
data(ames, package = "modeldata")

set.seed(123)
group_vfold_cv(ames, group = Neighborhood, v = 5)
#> # Group 5-fold cross-validation 
#> # A tibble: 5 × 2
#>   splits             id       
#>   <list>             <chr>    
#> 1 <split [2449/481]> Resample1
#> 2 <split [2642/288]> Resample2
#> 3 <split [2218/712]> Resample3
#> 4 <split [2367/563]> Resample4
#> 5 <split [2044/886]> Resample5
group_vfold_cv(
  ames,
  group = Neighborhood,
  v = 5,
  balance = "observations"
)
#> # Group 5-fold cross-validation 
#> # A tibble: 5 × 2
#>   splits             id       
#>   <list>             <chr>    
#> 1 <split [2348/582]> Resample1
#> 2 <split [2329/601]> Resample2
#> 3 <split [2347/583]> Resample3
#> 4 <split [2356/574]> Resample4
#> 5 <split [2340/590]> Resample5
```

<sup>Created on 2022-06-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

File: R/make_groups.R [hunk 3]
```diff
@@ -10,24 +10,26 @@
 #'  assessment set within a fold.
 #' @param balance If `v` is less than the number of unique groups, how should
 #'  groups be combined into folds? Should be one of
-#'  `groups`, `observations`.
+#'  `"groups"`, `"observations"`, `"prop"`.
```
[the remark this answers is not in the document]
Comment (mikemahoney218) [reply R/make_groups.R:13]: I attempted to address this in 8b1dccb -- let me know what you think

[the remark this answers is not in the document]
Comment (juliasilge) [reply R/make_groups.R:13]: Looks really good!

**Pull request merged**